File: python/library/aliyun/oss.py
from config.aliyun import Aliyun
from .signature import Signature
from util.util import Util
from util.hash import Hash
from util.base64 import Base64
import oss2

# 对象存储
class Oss:

  OssConn = None              #连接
  AccessKeyId: str = ''       #RAM: AccessKeyId
  AccessKeySecret: str = ''   #RAM: AccessKeySecret
  Endpoint: str = ''          #地域节点
  Bucket: str = ''            #Bucket名称

  # 签名直传
  def Policy(dir: str, file: str, expireTime: int=0, maxSize: int=0):
    ram = Aliyun.RAM()
    cfg = Aliyun.OSS()
    # 默认值
    if expireTime == 0 : expireTime = cfg['ExpireTime']
    if maxSize == 0 : maxSize = cfg['MaxSize']
    # 数据
    res = Signature.PolicySign(expireTime, maxSize)
    res['host'] = 'https://'+cfg['Bucket']+'.'+cfg['Endpoint']
    res['dir'] = dir
    res['file'] = file
    res['max_size'] = maxSize
    # 回调
    callbackBody = Util.JsonEncode({
      'dir': dir,
      'file': file,
      'expire': res['expire'],
      'sign': Hash.Md5(dir+'&'+file+'&'+str(res['expire'])+'&'+ram['AccessKeySecret']),
    })
    callbackData = Util.JsonEncode({
      'callbackUrl': cfg['CallbackUrl'],
      'callbackBodyType': cfg['CallbackType'],
      'callbackBody': callbackBody,
    })
    res['callback'] = Base64.ToStr(Base64.Encode(Base64.ToByte(callbackData)))
    return res

  # 签名直传-验证
  def PolicyVerify(dir: str, file: str, expire: str, sign: str):
    # 配置
    ram = Aliyun.RAM()
    # 验证
    signTmp = Hash.Md5(dir+'&'+file+'&'+expire+'&'+ram['AccessKeySecret'])
    if sign != signTmp : return False
    # 是否超时
    now = Util.Time()
    etime = int(expire)
    if now > etime : return False
    return True

  # 初始化
  def Init():
    # 配置
    ramCfg = Aliyun.RAM()
    ossCfg = Aliyun.OSS()
    if not Oss.AccessKeyId : Oss.AccessKeyId = ramCfg['AccessKeyId']
    if not Oss.AccessKeySecret : Oss.AccessKeySecret = ramCfg['AccessKeySecret']
    if not Oss.Endpoint : Oss.Endpoint = ossCfg['Endpoint']
    if not Oss.Bucket : Oss.Bucket = ossCfg['Bucket']
    # 连接
    if not Oss.OssConn :
      try:
        auth = oss2.Auth(Oss.AccessKeyId, Oss.AccessKeySecret)
        Oss.OssConn = oss2.Bucket(auth, Oss.Endpoint, Oss.Bucket)
      except Exception as e:
        print('[OSS] Conn:', e)
        Oss.OssConn = None
    return Oss.OssConn

  # 列表
  def ListObject(path: str):
    res = {'folder': [], 'file': []}
    # 连接
    conn = Oss.Init()
    if not conn : return res
    # 数据
    for val in oss2.ObjectIterator(conn, prefix = path, delimiter = '/'):
      if val.is_prefix(): res['folder'] += [val.key]
      else : res['file'] += [val.key]
    return res

  # 上传
  def PutObject(file: str, content, headers=None):
    # 连接
    conn = Oss.Init()
    if not conn : return False
    # 执行
    conn.put_object(file, content, headers)
    return True

  # 删除-单个
  def DeleteObject(file: str):
    if len(file)==0 : return False
    # 连接
    conn = Oss.Init()
    if not conn : return False
    # 执行
    conn.delete_object(file)
    return True

  # 删除-多个
  def DeleteObjects(files: list):
    if len(files)==0 : return False
    # 连接
    conn = Oss.Init()
    if not conn : return False
    # 执行
    conn.batch_delete_objects(files)
    return True

  # 删除-文件夹&文件
  def DeleteObjectAll(path: str):
    if len(path)==0 : return False
    # 连接
    conn = Oss.Init()
    if not conn : return False
    # 文件
    last = path[-1:]
    if last != '/' :
      conn.delete_object(path)
      return True
    # 文件夹
    objects = []
    lists = oss2.ObjectIterator(conn, prefix=path)
    for val in lists :
      objects += [val.key]
    return Oss.DeleteObjects(objects)

File: java/src/main/java/webmis/library/aliyun/Oss.java
package webmis.library.aliyun;

import java.io.ByteArrayInputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import com.alibaba.fastjson.JSONObject;
import com.aliyun.oss.OSS;
import com.aliyun.oss.OSSClientBuilder;
import com.aliyun.oss.model.DeleteObjectsRequest;
import com.aliyun.oss.model.DeleteObjectsResult;
import com.aliyun.oss.model.ListObjectsRequest;
import com.aliyun.oss.model.OSSObjectSummary;
import com.aliyun.oss.model.ObjectListing;

import webmis.config.Aliyun;
import webmis.util.Base64;
import webmis.util.Hash;
import webmis.util.Type;
import webmis.util.Util;

/* 对象存储 */
public class Oss extends Signature {

  public static OSS OssConn = null;           //连接
  public static String AccessKeyId = "";      //RAM: AccessKeyId
  public static String AccessKeySecret = "";  //RAM: AccessKeySecret
  public static String Endpoint = "";         //地域节点
  public static String Bucket = "";           //Bucket名称

  /* 签名直传 */
  static public JSONObject Policy(String dir, String file, long expireTime, long maxSize) {
    HashMap<String, Object> ram = Aliyun.RAM();
    HashMap<String, Object> cfg = Aliyun.OSS();
    // 默认值
    if(expireTime==0) expireTime = Type.Long(cfg.get("ExpireTime"));
    if(maxSize==0) maxSize = Type.Long(cfg.get("MaxSize"));
    // 数据
    JSONObject res = PolicySign(expireTime, maxSize);
    res.put("host", "https://"+cfg.get("Bucket").toString()+"."+cfg.get("Endpoint").toString());
    res.put("dir", dir);
    res.put("file", file);
    res.put("max_size", maxSize);
    // 回调
    JSONObject callbackBody = new JSONObject();
    callbackBody.put("dir", dir);
    callbackBody.put("file", file);
    callbackBody.put("expire", res.get("expire"));
    callbackBody.put("sign", Hash.Md5(dir+"&"+file+"&"+String.valueOf(res.get("expire"))+"&"+ram.get("AccessKeySecret").toString()));
    JSONObject callbackData = new JSONObject();
    callbackData.put("callbackUrl", cfg.get("CallbackUrl"));
    callbackData.put("callbackBodyType", cfg.get("CallbackType"));
    callbackData.put("callbackBody", Util.JsonEncode(callbackBody));
    res.put("callback", Base64.ToStr(Base64.Encode(Base64.ToByte(Util.JsonEncode(callbackData)))));
    return res;
  }

  /* 签名直传-验证 */
  public static Boolean PolicyVerify(String dir, String file, String expire, String sign) {
    // 配置
    HashMap<String, Object> ram = Aliyun.RAM();
    // 验证
    String signTmp = Hash.Md5(dir+"&"+file+"&"+expire+"&"+ram.get("AccessKeySecret").toString());
    if(!sign.equals(signTmp)) return false;
    // 是否超时
    long now = Util.Time();
    long etime = Long.valueOf(expire);
    if(now > etime) return false;
    return true;
  }

  /* 初始化 */
  public static OSS Init() {
    // 配置
    HashMap<String, Object> ramCfg = Aliyun.RAM();
    HashMap<String, Object> ossCfg = Aliyun.OSS();
    if(AccessKeyId=="") AccessKeyId = ramCfg.get("AccessKeyId").toString();
    if(AccessKeySecret=="") AccessKeySecret = ramCfg.get("AccessKeySecret").toString();
    if(Endpoint=="") Endpoint = ossCfg.get("Endpoint").toString();
    if(Bucket=="") Bucket = ossCfg.get("Bucket").toString();
    // 连接
    if(OssConn==null){
      OssConn = new OSSClientBuilder().build(Endpoint, AccessKeyId, AccessKeySecret);
    }
    return OssConn;
  }

  /* 列表 */
  static public HashMap<String, ArrayList<String>> ListObject(String path) {
    HashMap<String, ArrayList<String>> res = new HashMap<>();
    ArrayList<String> folder = new ArrayList<>();
    ArrayList<String> file = new ArrayList<>();
    res.put("folder", folder);
    res.put("file", file);
    // 连接
    OSS conn = Init();
    if(conn==null) return res;
    ObjectListing objectListing = conn.listObjects(new ListObjectsRequest(Bucket).withPrefix(path).withDelimiter("/"));
    // 文件夹
    for (String val : objectListing.getCommonPrefixes()) {
      folder.add(val);
    }
    // 文件
    List<OSSObjectSummary> sums = objectListing.getObjectSummaries();
    for (OSSObjectSummary s : sums) {
      file.add(s.getKey());
    }
    // 关闭
    conn.shutdown();
    return res;
  }

  /* 上传 */
  static public boolean PutObject(String file, byte[] content) {
    if(file.isEmpty()) return false;
    // 连接
    OSS conn = Init();
    if(conn==null) return false;
    // 执行
    conn.putObject(Bucket, file, new ByteArrayInputStream(content));
    return true;
  }

  /* 删除-单个 */
  static public boolean DeleteObject(String file) {
    if(file.isEmpty()) return false;
    // 连接
    OSS conn = Init();
    if(conn==null) return false;
    // 执行
    try{
      conn.deleteObject(Bucket, file);
      // 关闭
      conn.shutdown();
      return true;
    } catch (Exception e) {
      Print("[OSS] Dels: ", e.getMessage());
      return false;
    }
  }

  /* 删除-多个 */
  static public boolean DeleteObjects(ArrayList<String> files) {
    if(files==null) return false;
    // 连接
    OSS conn = Init();
    if(conn==null) return false;
    // 执行
    try{
      DeleteObjectsResult deleteObjectsResult = conn.deleteObjects(new DeleteObjectsRequest(Bucket).withKeys(files));
      deleteObjectsResult.getDeletedObjects();
      // 关闭
      conn.shutdown();
      return true;
    } catch (Exception e) {
      Print("[OSS] Dels: ", e.getMessage());
      return false;
    }
  }

  /* 删除-文件夹&文件 */
  static public boolean DeleteObjectAll(String path) {
    if(path.isEmpty()) return false;
    // 连接
    OSS conn = Init();
    if(conn==null) return false;
    try{
      // 文件
      String last = path.substring(path.length()-1, path.length());
      if(last != "/") conn.deleteObject(Bucket, path);
      // 文件夹
      String nextMarker = null;
      ObjectListing objectListing = null;
      ArrayList<String> objects = new ArrayList<String>();
      do {
        ListObjectsRequest listObjectsRequest = new ListObjectsRequest(Bucket).withPrefix(path).withMarker(nextMarker);
        objectListing = conn.listObjects(listObjectsRequest);
        for (OSSObjectSummary val : objectListing.getObjectSummaries()) {
          objects.add(val.getKey());
        }
      } while (objectListing.isTruncated());
      return DeleteObjects(objects);
    } catch (Exception e) {
      Print("[OSS] DelAll: ", e.getMessage());
      return false;
    }
  }
  
}
